### PR TITLE
finished learning rate system

### PR DESCRIPTION
**Pull request by vbvg2008**

### CODE REVIEW

File: fastestimator/trace/adapt/reduce_lr_on_plateau.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        if self.monitor_op(data[self.metric], self.best):
+            self.best = data[self.metric]
+            self.wait = 0
+        else:
+            self.wait += 1
+            if self.wait >= self.patience:
+                new_lr = max(self.min_lr, np.float32(self.factor * get_lr(self.model)))
+                set_lr(self.model, new_lr)
+                self.wait = 0
+                print("FastEstimator-ReduceLROnPlateau: learning rate reduced to {}".format(new_lr))
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/reduce_lr_on_plateau.py]: I think we want to put this into the logs since they're now integrated with out experiment summary / auditing stuff

File: fastestimator/trace/adapt/reduce_lr_on_plateau.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        min_lr: minimum learning rate, defaults to 1e-6.
+    """
+    system: System
+
+    def __init__(self,
+                 model: Union[tf.keras.Model, torch.nn.Module],
+                 metric: Optional[str] = None,
+                 patience: int = 10,
+                 factor: float = 0.1,
+                 reduce_mode: str = "min",
```
Comment (TortoiseHam) [inline fastestimator/trace/adapt/reduce_lr_on_plateau.py]: 'reduce_mode' name isn't super intuitive - at first it thought it might be controlling whether learning rate increases or decreases

Comment (vbvg2008) [reply fastestimator/trace/adapt/reduce_lr_on_plateau.py]: done

File: fastestimator/trace/adapt/reduce_lr_on_plateau.py (hunk 1 above)
Comment (vbvg2008) [reply fastestimator/trace/adapt/reduce_lr_on_plateau.py]: done

**Comment by TortoiseHam** [comment]
can we add learning rate logging to the train essential on start for sake of log visualization? other than that good to go

File: fastestimator/schedule/lr_shedule.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+def cosine_decay(epoch_or_step: int, cycle_length: int, init_lr: float, min_lr: float = 1e-6, start: int = 0):
+    """ learning rate cosine decay function (using half of cosine curve)
+
+    Args:
+        epoch_or_step: current step or epoch during training.
+        cycle_length: the decay cycle length.
+        init_lr: initial learning rate to decay from.
+        min_lr: minimum learning rate.
+        start_step: the step or epoch to start the decay schedule.
```
Comment (TortoiseHam) [inline fastestimator/schedule/lr_shedule.py]: start_step vs start in method sig

File: fastestimator/schedule/lr_shedule.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import math
+
+
+def cosine_decay(epoch_or_step: int, cycle_length: int, init_lr: float, min_lr: float = 1e-6, start: int = 0):
```
Comment (TortoiseHam) [inline fastestimator/schedule/lr_shedule.py]: maybe just 'time' instead of 'epoch_or_step'

File: fastestimator/trace/trace.py [hunk 5]
```diff
@@ -97,6 +97,9 @@ def on_epoch_end(self, data: Data):
 
     def on_end(self, data: Data):
         data.write_with_log("total_time", "{} sec".format(round(time.perf_counter() - self.train_start, 2)))
+        for model in self.system.network.models:
```
Comment (TortoiseHam) [inline fastestimator/trace/trace.py]: on begin

File: fastestimator/schedule/lr_shedule.py (hunk 3 above)
Comment (vbvg2008) [reply fastestimator/schedule/lr_shedule.py]: done

File: fastestimator/schedule/lr_shedule.py (hunk 4 above)
Comment (vbvg2008) [reply fastestimator/schedule/lr_shedule.py]: done

File: fastestimator/trace/trace.py (hunk 5 above)
Comment (vbvg2008) [reply fastestimator/trace/trace.py]: done
